File: backend/ingestion/src/domain/repositories.ts
/**
 * Repository interfaces for data access
 * These define the contract for database operations
 */

import {
  Account,
  Category,
  ImportBatch,
  Transaction,
  Rule,
  ClassificationOverride,
  CategoryTotalsView,
  CreateAccountInput,
  CreateCategoryInput,
  CreateImportBatchInput,
  CreateTransactionInput,
  CreateRuleInput,
  CreateClassificationOverrideInput,
  AccountStatus,
} from './types';

/**
 * AccountRepository - Data access for Account entities
 */
export interface IAccountRepository {
  /**
   * Find account by ID
   */
  findById(id: string): Promise<Account | null>;

  /**
   * Find all accounts for a user/organization
   */
  findAll(status?: AccountStatus): Promise<Account[]>;

  /**
   * Create a new account
   */
  create(input: CreateAccountInput): Promise<Account>;

  /**
   * Update existing account
   */
  update(id: string, updates: Partial<CreateAccountInput>): Promise<Account>;

  /**
   * Delete (archive) account
   */
  delete(id: string): Promise<void>;

  /**
   * Count accounts with specific status
   */
  count(status?: AccountStatus): Promise<number>;
}

/**
 * CategoryRepository - Data access for Category entities
 */
export interface ICategoryRepository {
  /**
   * Find category by ID
   */
  findById(id: string): Promise<Category | null>;

  /**
   * Find category by name
   */
  findByName(name: string): Promise<Category | null>;

  /**
   * Find all categories
   */
  findAll(): Promise<Category[]>;

  /**
   * Find categories by type (RECEITA or DESPESA)
   */
  findByType(tipo: 'RECEITA' | 'DESPESA'): Promise<Category[]>;

  /**
   * Create a new category
   */
  create(input: CreateCategoryInput): Promise<Category>;

  /**
   * Update existing category
   */
  update(id: string, updates: Partial<CreateCategoryInput>): Promise<Category>;

  /**
   * Delete category
   */
  delete(id: string): Promise<void>;

  /**
   * Count all categories
   */
  count(): Promise<number>;
}

/**
 * ImportBatchRepository - Data access for ImportBatch entities
 */
export interface IImportBatchRepository {
  /**
   * Find batch by ID
   */
  findById(id: string): Promise<ImportBatch | null>;

  /**
   * Find batch by checksum with account and period (duplicate detection)
   */
  findByChecksum(
    accountId: string,
    fileChecksum: string,
    periodMonth: number,
    periodYear: number
  ): Promise<ImportBatch | null>;

  /**
   * Find all batches for an account
   */
  findByAccountId(
    accountId: string,
    periodMonth?: number,
    periodYear?: number
  ): Promise<ImportBatch[]>;

  /**
   * Find batches by account (paginated)
   */
  findAllByAccountId(
    accountId: string,
    limit?: number,
    offset?: number
  ): Promise<{ batches: ImportBatch[]; total: number }>;

  /**
   * Create a new batch
   */
  create(input: CreateImportBatchInput): Promise<ImportBatch>;

  /**
   * Update row count after import
   */
  updateRowCount(id: string, rowCount: number): Promise<ImportBatch>;

  /**
   * Get batch statistics
   */
  getStats?(accountId: string): Promise<{
    totalBatches: number;
    totalTransactions: number;
    dateRange: { from: Date; to: Date };
  }>;
}

/**
 * TransactionRepository - Data access for Transaction entities
 */
export interface ITransactionRepository {
  /**
   * Find transaction by ID
   */
  findById(id: string): Promise<Transaction | null>;

  /**
   * Find all transactions in a batch
   */
  findByBatchId(batchId: string): Promise<Transaction[]>;

  /**
   * Find transactions by batch and classification status
   */
  findByBatchAndStatus(
    batchId: string,
    status: 'classified' | 'unclassified'
  ): Promise<Transaction[]>;

  /**
   * Find unclassified transactions for an account (paginated)
   */
  findUnclassified(
    accountId: string,
    limit?: number,
    offset?: number
  ): Promise<{ transactions: Transaction[]; total: number }>;

  /**
   * Find transactions by account (paginated)
   */
  findByAccountId(
    accountId: string,
    limit?: number,
    offset?: number
  ): Promise<{ transactions: Transaction[]; total: number }>;

  /**
   * Create a new transaction
   */
  create(input: CreateTransactionInput): Promise<Transaction>;

  /**
   * Create multiple transactions (batch insert)
   */
  createMany(inputs: CreateTransactionInput[]): Promise<Transaction[]>;

  /**
   * Update transaction classification
   */
  update(
    id: string,
    updates: Partial<{
      categoryId: string;
      tipo: 'RECEITA' | 'DESPESA';
      classificationSource: ClassificationSource;
      ruleId: string;
      ruleVersion: number;
      rationale: string;
    }>
  ): Promise<Transaction>;

  /**
   * Count transactions in a batch
   */
  countByBatchId(batchId: string): Promise<number>;

  /**
   * Get classification stats for a batch
   */
  getClassificationStats(batchId: string): Promise<{
    classified: number;
    unclassified: number;
  }>;
}

/**
 * RuleRepository - Data access for Rule entities
 */
export interface IRuleRepository {
  /**
   * Find rule by ID
   */
  findById(id: string): Promise<Rule | null>;

  /**
   * Find all active rules
   */
  findActive(): Promise<Rule[]>;

  /**
   * Find rules for a specific category
   */
  findByCategory(categoryId: string): Promise<Rule[]>;

  /**
   * Find rules by type (RECEITA or DESPESA)
   */
  findByType(tipo: 'RECEITA' | 'DESPESA'): Promise<Rule[]>;

  /**
   * Create a new rule
   */
  create(input: CreateRuleInput): Promise<Rule>;

  /**
   * Update rule (increments version)
   */
  update(id: string, updates: Partial<Omit<CreateRuleInput, 'createdBy'>>): Promise<Rule>;

  /**
   * Deactivate rule (soft delete)
   */
  deactivate(id: string): Promise<void>;

  /**
   * Count active rules
   */
  countActive(): Promise<number>;
}

/**
 * ClassificationOverrideRepository - Data access for ClassificationOverride entities
 */
export interface IClassificationOverrideRepository {
  /**
   * Find override by ID
   */
  findById(id: string): Promise<ClassificationOverride | null>;

  /**
   * Find override for a transaction
   */
  findByTransaction(transactionId: string): Promise<ClassificationOverride | null>;

  /**
   * Find all overrides by actor (user)
   */
  findByActor(actor: string): Promise<ClassificationOverride[]>;

  /**
   * Find overrides in date range
   */
  findByDateRange(startDate: Date, endDate: Date): Promise<ClassificationOverride[]>;

  /**
   * Create a new override
   */
  create(input: CreateClassificationOverrideInput): Promise<ClassificationOverride>;

  /**
   * Get override statistics
   */
  getStats(
    startDate?: Date,
    endDate?: Date
  ): Promise<{
    total: number;
    byActor: Record<string, number>;
    byCategory: Record<string, number>;
  }>;
}

/**
 * ReportingRepository - Data access for materialized views
 */
export interface IReportingRepository {
  /**
   * Get category totals for reporting
   */
  getCategoryTotals(
    year?: number,
    month?: number,
    accountId?: string
  ): Promise<CategoryTotalsView[]>;

  /**
   * Get category totals for a specific account
   */
  getCategoryTotalsByAccount(accountId: string, year?: number): Promise<CategoryTotalsView[]>;

  /**
   * Refresh materialized view
   * Should be called after significant data imports
   */
  refreshCategoryTotals(): Promise<void>;

  /**
   * Get summary statistics
   */
  getSummaryStats(accountId?: string): Promise<{
    totalTransactions: number;
    totalAmount: number;
    averageAmount: number;
    byCategory: Array<{ category: string; total: number; count: number }>;
  }>;
}

File: backend/reporting/src/main/kotlin/org/example/financeiro/reporting/domain/Repositories.kt
package org.example.financeiro.reporting.domain

import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository
import java.time.LocalDate
import java.time.LocalDateTime

/**
 * Spring Data JPA Repository for Account entities
 */
@Repository
interface AccountRepository : JpaRepository<Account, String> {
    fun findByStatus(status: AccountStatus): List<Account>
    fun countByStatus(status: AccountStatus): Long
}

/**
 * Spring Data JPA Repository for Category entities
 */
@Repository
interface CategoryRepository : JpaRepository<Category, String> {
    fun findByName(name: String): Category?
    fun findByTipo(tipo: TransactionType): List<Category>
    fun countByTipo(tipo: TransactionType): Long
}

/**
 * Spring Data JPA Repository for ImportBatch entities
 */
@Repository
interface ImportBatchRepository : JpaRepository<ImportBatch, String> {
    fun findByAccount(account: Account): List<ImportBatch>

    @Query("SELECT b FROM ImportBatch b WHERE b.account.id = :accountId AND b.fileChecksum = :checksum")
    fun findByAccountAndChecksum(
        @Param("accountId") accountId: String,
        @Param("checksum") checksum: String
    ): ImportBatch?

    @Query("SELECT b FROM ImportBatch b WHERE b.periodYear = :year AND b.periodMonth = :month")
    fun findByPeriod(
        @Param("year") year: Int,
        @Param("month") month: Int
    ): List<ImportBatch>

    @Query("""
        SELECT NEW map(
            COUNT(b) as totalBatches,
            COALESCE(SUM(b.rowCount), 0) as totalTransactions,
            MIN(b.uploadedAt) as minDate,
            MAX(b.uploadedAt) as maxDate
        )
        FROM ImportBatch b
        WHERE b.account.id = :accountId
    """)
    fun getStatsByAccount(@Param("accountId") accountId: String): Map<String, Any>?
}

/**
 * Spring Data JPA Repository for Transaction entities
 */
@Repository
interface TransactionRepository : JpaRepository<Transaction, String> {
    fun findByBatch(batch: ImportBatch): List<Transaction>

    @Query("SELECT t FROM Transaction t WHERE t.account.id = :accountId AND t.category IS NULL")
    fun findUnclassified(@Param("accountId") accountId: String?): List<Transaction>

    @Query("""
        SELECT t FROM Transaction t
        WHERE t.date >= :startDate AND t.date <= :endDate
        AND (:accountId IS NULL OR t.account.id = :accountId)
    """)
    fun findByDateRange(
        @Param("startDate") startDate: LocalDate,
        @Param("endDate") endDate: LocalDate,
        @Param("accountId") accountId: String?
    ): List<Transaction>

    fun findByCategory(category: Category): List<Transaction>

    @Query("SELECT COUNT(t) FROM Transaction t WHERE t.category IS NULL AND (:accountId IS NULL OR t.account.id = :accountId)")
    fun countUnclassified(@Param("accountId") accountId: String?): Long

    @Query("""
        SELECT NEW map(
            COUNT(t) as total,
            COUNT(CASE WHEN t.category IS NOT NULL THEN 1 END) as classified,
            COUNT(CASE WHEN t.category IS NULL THEN 1 END) as unclassified
        )
        FROM Transaction t
        WHERE (:accountId IS NULL OR t.account.id = :accountId)
        AND (:startDate IS NULL OR t.date >= :startDate)
        AND (:endDate IS NULL OR t.date <= :endDate)
    """)
    fun getStatistics(
        @Param("accountId") accountId: String?,
        @Param("startDate") startDate: LocalDate?,
        @Param("endDate") endDate: LocalDate?
    ): Map<String, Any>?
}

/**
 * Spring Data JPA Repository for Rule entities
 */
@Repository
interface RuleRepository : JpaRepository<Rule, String> {
    fun findByActiveTrue(): List<Rule>
    fun findByCategory(category: Category): List<Rule>
    fun findByTipo(tipo: TransactionType): List<Rule>
    fun findByActiveTrueAndTipo(tipo: TransactionType): List<Rule>
    fun countByActiveTrue(): Long
}

/**
 * Spring Data JPA Repository for ClassificationOverride entities
 */
@Repository
interface ClassificationOverrideRepository : JpaRepository<ClassificationOverride, String> {
    fun findByTransaction(transaction: Transaction): ClassificationOverride?

    @Query("SELECT o FROM ClassificationOverride o WHERE o.actor = :actor")
    fun findByActor(@Param("actor") actor: String): List<ClassificationOverride>

    @Query("""
        SELECT o FROM ClassificationOverride o
        WHERE o.createdAt >= :startDate AND o.createdAt <= :endDate
    """)
    fun findByDateRange(
        @Param("startDate") startDate: LocalDateTime,
        @Param("endDate") endDate: LocalDateTime
    ): List<ClassificationOverride>

    @Query("""
        SELECT NEW map(
            COUNT(o) as total,
            COUNT(DISTINCT o.actor) as uniqueActors,
            COUNT(DISTINCT o.newCategory.id) as uniqueCategories
        )
        FROM ClassificationOverride o
        WHERE (:startDate IS NULL OR o.createdAt >= :startDate)
        AND (:endDate IS NULL OR o.createdAt <= :endDate)
    """)
    fun getStatistics(
        @Param("startDate") startDate: LocalDateTime?,
        @Param("endDate") endDate: LocalDateTime?
    ): Map<String, Any>?
}

/**
 * Spring Data JPA Repository for MaterializedView (CategoryTotals)
 * Read-only repository for reporting queries
 */
@Repository
interface CategoryTotalsViewRepository : JpaRepository<CategoryTotalsView, Long> {
    @Query("""
        SELECT v FROM CategoryTotalsView v
        WHERE (:year IS NULL OR v.year = :year)
        AND (:month IS NULL OR v.month = :month)
        AND (:accountId IS NULL OR v.accountId = :accountId)
    """)
    fun findTotals(
        @Param("year") year: Int?,
        @Param("month") month: Int?,
        @Param("accountId") accountId: String?
    ): List<CategoryTotalsView>

    @Query("SELECT v FROM CategoryTotalsView v WHERE v.accountId = :accountId AND (:year IS NULL OR v.year = :year)")
    fun findByAccount(
        @Param("accountId") accountId: String,
        @Param("year") year: Int?
    ): List<CategoryTotalsView>

    @Query("""
        SELECT NEW map(
            SUM(v.totalAmount) as totalAmount,
            SUM(v.rowCount) as totalRows,
            COUNT(DISTINCT v.categoryId) as categoryCount
        )
        FROM CategoryTotalsView v
        WHERE v.accountId = :accountId
        AND (:year IS NULL OR v.year = :year)
    """)
    fun getSummary(
        @Param("accountId") accountId: String,
        @Param("year") year: Int?
    ): Map<String, Any>?
}
